cmd/checkheader/main.go

```diff
@@ -197,108 +196,74 @@ func tryConvertGoConst(w io.Writer, l string) {
 // the same value of i must not be passed again in a future call.
 // The value of i should be generated by a counter.
 func tryConvertDefineFunc(w io.Writer, l string, i int) bool {
-	var versionCond string
-	if strings.HasPrefix(l, "/*check:from=") {
-		i1 := strings.Index(l, "=")
-		i2 := strings.Index(l, "*/")
-		if i1 < 0 || i2 < 0 {
-			log.Fatalln("unexpected 'check:from' condition: " + l)
-		}
-		from := l[i1+1 : i2]
-		switch from {
-		case "1.1.0":
-			versionCond = "OPENSSL_VERSION_NUMBER >= 0x10100000L"
-		case "1.1.1":
-			versionCond = "OPENSSL_VERSION_NUMBER >= 0x10101000L"
-		case "3.0.0":
-			versionCond = "OPENSSL_VERSION_NUMBER >= 0x30000000L"
-		default:
-			log.Println("unexpected 'check:from' version" + l)
-			return false
-		}
-		if l[i2+2] != ' ' {
-			log.Fatalln("missing space between 'check:from' condition and function macro: " + l)
-		}
-		l = l[i2+3:]
-	}
-	if !strings.HasPrefix(l, "DEFINEFUNC") {
+	if !strings.HasPrefix(l, "/*[[mkcgo") {
 		return false
 	}
-	if strings.HasPrefix(l, "DEFINEFUNC_VARIADIC") {
+	if strings.Contains(l, "[[mkcgo::variadic(") {
 		// Variadic functions are not supported. There is not enough
-		// information in the macro to create use it in writeDefineFunc.
+		// information to create use it in writeDefineFunc.
 		return false
 	}
-	i1 := strings.IndexByte(l, '(')
-	// The first ")," match is always the end of the argument list parameter.
-	// We are not interested in the last parameter and parsing them would complicate the algorithm.
-	// Matching against ')' is not enough as it also appears when the argument list parameter contains function pointers.
-	i2 := strings.Index(l, "),")
+	i1 := strings.LastIndex(l, "*/ ")
+	i2 := strings.LastIndexByte(l, ')')
 	if i1 < 0 || i2 < 0 {
 		log.Println("unexpected argument list in function line: " + l)
 		return false
 	}
-	subs := l[i1+1 : i2+1]
+	subs := l[i1+3 : i2+1]
 	writeCheck := func(ret, name, args string) {
-		fmt.Fprintf(w, "%s(*__check_%d)%s = %s;\n", ret, i, args, name)
+		fmt.Fprintf(w, "%s(*__check_%d)(%s) = %s;\n", ret, i, args, name)
 	}
 	writeDefineFunc := func(cond string) {
-		args := strings.SplitN(subs, ",", 3)
-		if len(args) < 3 {
-			log.Printf("wrong number of function macro arguments in line: %s\n", l)
+		argsAndName, subs, ok := strings.Cut(subs, "(")
+		if !ok {
+			log.Printf("can't retrive return function name and return type in line: %s\n", l)
 			return
 		}
-		fnret, fnname, fnargs := args[0], args[1], args[2]
+		idx := strings.LastIndexByte(argsAndName, ' ')
+		if idx < 0 {
+			log.Printf("can't retrive return type in line: %s\n", l)
+			return
+		}
+		fnret := argsAndName[:idx]
+		fnname := argsAndName[idx+1:]
```

Comment:
Would it make sense to share the parser with `mkcgo`? (Index-based logic in particular usually seems fiddly and panic-happy to me, enough to be worth sharing to at least reduce the amount of it.)

Reply:
Good suggestion. On the other hand, I plan to remove the checkheader program in a follow-up PR, and teach `mkcgo` to generate the check header. So I would leave it as is for now to not make this PR even more complex.